clang/test/SemaSYCL/intel-fpga-no-global-work-offset.cpp

```diff
@@ -1,51 +1,47 @@
+
+#include "sycl.hpp"
+
+using namespace cl::sycl;
+queue q;
 
 struct FuncObj {
   //expected-warning@+2 {{attribute 'intelfpga::no_global_work_offset' is deprecated}}
   //expected-note@+1 {{did you mean to use 'intel::no_global_work_offset' instead?}}
-  [[intelfpga::no_global_work_offset]] void operator()() {}
+  [[intelfpga::no_global_work_offset]] void operator()() const {}
 };
 
-template <typename name, typename Func>
-void kernel(Func kernelFunc) {
-  kernelFunc();
-}
-
 int main() {
-  // CHECK: SYCLIntelNoGlobalWorkOffsetAttr{{.*}}Enabled
-  kernel<class test_kernel1>([]() {
-    FuncObj();
-  });
-
-  // CHECK: SYCLIntelNoGlobalWorkOffsetAttr
-  // CHECK-NOT: Enabled
-  kernel<class test_kernel2>(
-      []() [[intel::no_global_work_offset(0)]]{});
-
-  // CHECK: SYCLIntelNoGlobalWorkOffsetAttr{{.*}}Enabled
-  // expected-warning@+2{{'no_global_work_offset' attribute should be 0 or 1. Adjusted to 1}}
-  kernel<class test_kernel3>(
-      []() [[intel::no_global_work_offset(42)]]{});
-
-  // expected-error@+2{{'no_global_work_offset' attribute requires a non-negative integral compile time constant expression}}
-  kernel<class test_kernel4>(
-      []() [[intel::no_global_work_offset(-1)]]{});
-
-  // expected-error@+2{{'no_global_work_offset' attribute requires parameter 0 to be an integer constant}}
-  kernel<class test_kernel5>(
-      []() [[intel::no_global_work_offset("foo")]]{});
-
-  kernel<class test_kernel6>([]() {
-    // expected-error@+1{{'no_global_work_offset' attribute only applies to functions}}
-    [[intel::no_global_work_offset(1)]] int a;
+  q.submit([&](handler &h) {
+    // CHECK:       SYCLIntelNoGlobalWorkOffsetAttr {{.*}}
+    // CHECK-NEXT:  IntegerLiteral{{.*}}1{{$}}
+    h.single_task<class test_kernel1>(FuncObj());
+
+    // CHECK: SYCLIntelNoGlobalWorkOffsetAttr {{.*}}
+    // CHECK-NEXT:  IntegerLiteral{{.*}}0{{$}}
+    h.single_task<class test_kernel2>(
+        []() [[intel::no_global_work_offset(0)]]{});
+
+    // CHECK: SYCLIntelNoGlobalWorkOffsetAttr{{.*}}
+    h.single_task<class test_kernel3>(
+        []() [[intel::no_global_work_offset(42)]]{});
+
+    // expected-error@+2{{'no_global_work_offset' attribute requires a non-negative integral compile time constant expression}}
+    h.single_task<class test_kernel4>(
+        []() [[intel::no_global_work_offset(-1)]]{});
+
+    // expected-error@+2{{'no_global_work_offset' attribute requires an integer constant}}
+    h.single_task<class test_kernel5>(
+        []() [[intel::no_global_work_offset("foo")]]{});
+
+    h.single_task<class test_kernel6>([]() {
+      // expected-error@+1{{'no_global_work_offset' attribute only applies to functions}}
+      [[intel::no_global_work_offset(1)]] int a;
+    });
+
+    // expected-warning@+2{{attribute 'no_global_work_offset' is already applied}}
+    h.single_task<class test_kernel7>(
+        []() [[intel::no_global_work_offset(0), intel::no_global_work_offset(1)]]{});
   });
```

Comment:
Additionally, you should add a test where a template instantiation is just plain wrong:
```
template <typename Ty>
[[intel::no_global_work_offset(Ty{})]] void func() {}

struct S {};
func<S>();
```
as well as a test where the expression is not a constant expression:
```
int foo();
[[intel::no_global_work_offset(foo() + 12)]] void func();
```
and a test where it is a constant expression:
```
constexpr int foo() { return 0; }
[[intel::no_global_work_offset(foo() + 12)]] void func();
```

Reply:
Done. I have added all test cases above inside clang/test/SemaSYCL/sycl-device-intel-fpga-no-global-work-offset.cpp.